ktor-client/ktor-client-core/common/src/io/ktor/client/features/HttpPlainText.kt

```diff
@@ -13,51 +13,165 @@ import kotlinx.io.core.*
 
 /**
  * [HttpClient] feature that encodes [String] request bodies to [TextContent]
- * using a specific charset defined at [HttpPlainText.defaultCharset].
- * And also processes the response body as [String].
- * @param defaultCharset: default charset to use. [Charsets.UTF_8] by default.
+ * and processes the response body as [String].
  *
- * NOTE: the [HttpPlainText.defaultCharset] is the default one for your JVM so can change between servers!
- *       So please, specify one if you want consistent results in all your deployments.
+ * To configure charsets set following properties in [HttpPlainText.Config].
  */
-class HttpPlainText(var defaultCharset: Charset) {
+class HttpPlainText(
+    acceptCharsets: Set<Charset>,
+    charsetQuality: Map<Charset, Float>,
+    sendCharset: Charset?,
+    internal val responseCharsetFallback: Charset
+) {
+    private val requestCharset: Charset
+    private val acceptCharsetHeader: String
 
-    internal fun read(call: HttpClientCall, body: Input): String {
-        val actualCharset = call.response.charset() ?: call.responseConfig.defaultCharset
-        return body.readText(charset = actualCharset)
+    init {
+        val withQuality = charsetQuality.toList().sortedByDescending { it.second }
+        val withoutQuality = acceptCharsets.filter { !charsetQuality.containsKey(it) }.sortedBy { it.name }
+
+        acceptCharsetHeader = buildString {
+            withoutQuality.forEach {
+                if (length > 0) append(",")
+                append(it.name)
+            }
+
+            withQuality.forEach { (charset, quality) ->
+                if (length > 0) append(",")
+
+                check(quality in 0.0..1.0)
+                append("${charset.name};q=$quality")
+            }
+
+            if (isEmpty()) {
+                append(responseCharsetFallback.name)
+            }
+        }
+
+        requestCharset = sendCharset
+            ?: withoutQuality.firstOrNull()
+                ?: withQuality.firstOrNull()?.first
+                ?: Charsets.UTF_8
     }
 
     /**
-     * [HttpPlainText] configuration.
+     * Charset configuration for [HttpPlainText] feature.
      */
     class Config {
+        internal val acceptCharsets: MutableSet<Charset> = mutableSetOf()
+        internal val charsetQuality: MutableMap<Charset, Float> = mutableMapOf()
+
+        /**
+         * Add [charset] to allowed list with selected [quality].
+         */
+        fun accept(charset: Charset, quality: Float? = null) {
+            acceptCharsets.add(charset)
+
+            if (quality == null) {
+                charsetQuality.remove(charset)
+            } else {
+                charsetQuality[charset] = quality
+            }
+        }
+
+        /**
+         * Explicit [Charset] for sending content.
+         *
+         * Send best from [accept] charset if null.
+         */
+        var sendCharset: Charset? = null
```

Comment:
> Send best from [accept] charset if null.

This is not exactly true: it may lead to an attempt to use inappropriate charset to encode a request that may cause random unmappable character errors because _best_ = "the first charset from the list of charsets with no quality specified or with the largest quality". So depending on charset registration order it may always work (if UTF8 was registered first) or sometimes fail depending on request's content. Such instability could cause confusion.